text/0000-post-installation-mirror.md

```diff
@@ -0,0 +1,124 @@
+
+- When installation a post install artifact from offline mirror
+  - extract the tar.gz file in place
+  - do not run install scripts
+  - install dependencies and create bin link as usual
+
+# How We Teach This
+
+*What names and terminology work best for these concepts and why?*
+"post installation artifacts" to distinguish from node modules
+"network restricted environments (NRE)" where access to Internet is controlled
+"hermetic builds" means all dependencies are included, could be used as synonym for "deterministic builds"
+
+*How is this idea best presented?*
+
+- Emphases on the concept of "hermetic builds".
+- Emphases on `yarn` usage within "network restricted environments (NRE)".
+
+*Would the acceptance of this proposal mean the Yarn documentation must be re-organized or altered?*
+
+No.
+
+*Does it change how Yarn is taught to new users at any level?*
+
+This feature should be considered as advanced and should be taught specifically to following users:
+- users who need to operate within network restricted environments (NRE)
+- users who value deterministic dependency resolution and hermetic builds above convenience
+
+*How should this feature be introduced and taught to existing Yarn users?*
+Explain the intended use case with illustrated work flow.
+
+# Drawbacks
+
+- Somewhat a deviation from current node community norm.  This feature may potentially require additional teaching.
+- Added complexity for installation work flow and offline mirror structure
+- Depending on directory structure to identify post installation modules
+
+The following points are drawbacks under certain circumstances and advantages for other circumstances.
+
+- Native extension support
+If a node module has native extensions, a stored post installation module will not work on platforms different from where the module is created.
+
+Currently, native extensions are compiled during installation. However, since the compiler and libraries are provided by running environment, the compilation output are not guaranteed to be repeatable. To ensure hermetic and repeatable build, a separate RFC is necessary due to the complexity of supporting node native extensions .
+
+- Lost of some flexibility
+In at least one package (cldr-data), [build result can be altered by environment variables $CLDR_COVERAGE](https://github.com/rxaviers/cldr-data-npm/blob/master/install.js#L91).  Caching post installation artifacts will lose this flexibility.
+
+- Update to installation time downloads are ignored / require explicit action
+Installation scripts tends to download the latest version of dependencies.  A stored post installation artifacts will always have the same version of dependencies and thus potentially will not have the latest dependencies.  To update such installation time downloaded dependencies, explicit actions from offline mirror maintainers will be required.
+
+# Alternatives

```

Comment:
As I said above, this RFC goes beyond the concerns of Offline Mirror feature.

I think the problem may be solved by caching and sharing a built package in some way.
This may not work across platforms and machines, depends on every package and how a project is built.
I would try:
- disable `lifecycle scripts` for a package that needs Internet (maybe have this setting in package.json)
- before package is installed from Yarn cache, replace the cache with the prebuilt content. Packing, sharing and replacing in cache could be automated in some way by Yarn or a plugin or a third party script